File: backend/internal/store/projects.go
package store

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"path/filepath"
	"strings"

	"github.com/xiaobei/codex-issue-runner/backend/internal/agent"
)

func (s *Store) ListProjects(ctx context.Context) ([]Project, error) {
	rows, err := s.db.QueryContext(ctx, projectSelect+` order by p.sort_order asc, p.created_at asc, p.id asc`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	projects := []Project{}
	for rows.Next() {
		p, err := scanProject(rows)
		if err != nil {
			return nil, err
		}
		projects = append(projects, p)
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}
	return s.attachProjectAgentProfiles(ctx, projects)
}

func (s *Store) CreateProject(ctx context.Context, p Project) (Project, error) {
	t := now()
	applyProjectDefaults(&p)
	if p.ApprovalPolicy == "" {
		p.ApprovalPolicy = "never"
	}
	if p.Sandbox == "" {
		p.Sandbox = "workspace-write"
	}
	nextOrder, err := s.nextProjectSortOrder(ctx)
	if err != nil {
		return Project{}, err
	}
	p.SortOrder = nextOrder
	_, err = s.db.ExecContext(ctx, `insert into projects
		(id, name, cwd, provider, provider_config_json, auto_run, model,
		approval_policy, sandbox, default_agent_profile_id, sort_order, created_at, updated_at)
		values (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		p.ID, p.Name, p.CWD, p.Provider, p.ProviderConfig, p.AutoRun, p.Model,
		p.ApprovalPolicy, p.Sandbox, p.DefaultAgentProfileID, p.SortOrder, t, t)
	if err != nil {
		return Project{}, err
	}
	return s.GetProject(ctx, p.ID)
}

func (s *Store) GetProject(ctx context.Context, id string) (Project, error) {
	row := s.db.QueryRowContext(ctx, projectSelect+` where p.id = ?`, id)
	p, err := scanProject(row)
	if errors.Is(err, sql.ErrNoRows) {
		return Project{}, ErrNotFound
	}
	if err != nil {
		return Project{}, err
	}
	return s.attachProjectAgentProfile(ctx, p)
}

func (s *Store) UpdateProject(ctx context.Context, id string, patch ProjectPatch) (Project, error) {
	p, err := s.GetProject(ctx, id)
	if err != nil {
		return Project{}, err
	}
	patch = applyProjectPatchDefaults(&p, patch)
	applyProjectPatch(&p, patch)
	_, err = s.db.ExecContext(ctx, `update projects set name=?, cwd=?, provider=?,
		provider_config_json=?, auto_run=?, model=?, approval_policy=?, sandbox=?,
		default_agent_profile_id=?, updated_at=? where id=?`,
		p.Name, p.CWD, p.Provider, p.ProviderConfig, p.AutoRun,
		p.Model, p.ApprovalPolicy, p.Sandbox, p.DefaultAgentProfileID, now(), id)
	if err != nil {
		return Project{}, err
	}
	return s.GetProject(ctx, id)
}

func (s *Store) ReorderProjects(ctx context.Context, ids []string) ([]Project, error) {
	if len(ids) == 0 {
		return nil, fmt.Errorf("project order 不能为空")
	}
	tx, err := s.db.BeginTx(ctx, nil)
	if err != nil {
		return nil, err
	}
	defer tx.Rollback()

	if err := validateProjectIDs(ctx, tx, ids); err != nil {
		return nil, err
	}
	updatedAt := now()
	for index, id := range ids {
		if _, err := tx.ExecContext(ctx, `update projects set sort_order=?, updated_at=? where id=?`,
			index+1, updatedAt, id); err != nil {
			return nil, err
		}
	}
	if err := tx.Commit(); err != nil {
		return nil, err
	}
	return s.ListProjects(ctx)
}

func (s *Store) attachProjectAgentProfile(ctx context.Context, project Project) (Project, error) {
	projects, err := s.attachProjectAgentProfiles(ctx, []Project{project})
	if err != nil || len(projects) == 0 {
		return project, err
	}
	return projects[0], nil
}

func (s *Store) attachProjectAgentProfiles(ctx context.Context, projects []Project) ([]Project, error) {
	for idx := range projects {
		profileID := strings.TrimSpace(projects[idx].DefaultAgentProfileID)
		if profileID == "" {
			continue
		}
		profile, err := s.GetAgentProfile(ctx, profileID)
		if errors.Is(err, ErrNotFound) {
			continue
		}
		if err != nil {
			return nil, err
		}
		projects[idx].DefaultAgentProfile = &profile
	}
	return projects, nil
}

func validateProjectIDs(ctx context.Context, tx *sql.Tx, ids []string) error {
	seen := make(map[string]bool, len(ids))
	for _, id := range ids {
		if strings.TrimSpace(id) == "" {
			return fmt.Errorf("project id 不能为空")
		}
		if seen[id] {
			return fmt.Errorf("project id 重复: %s", id)
		}
		seen[id] = true
	}

	var count int
	if err := tx.QueryRowContext(ctx, `select count(*) from projects`).Scan(&count); err != nil {
		return err
	}
	if count != len(ids) {
		return fmt.Errorf("project order 必须包含全部项目")
	}

	rows, err := tx.QueryContext(ctx, `select id from projects`)
	if err != nil {
		return err
	}
	defer rows.Close()
	for rows.Next() {
		var id string
		if err := rows.Scan(&id); err != nil {
			return err
		}
		if !seen[id] {
			return fmt.Errorf("project order 缺少项目: %s", id)
		}
	}
	return rows.Err()
}

func (s *Store) DeleteProject(ctx context.Context, id string) error {
	res, err := s.db.ExecContext(ctx, `delete from projects where id = ?`, id)
	if err != nil {
		return err
	}
	return requireAffected(res)
}

func applyProjectDefaults(p *Project) {
	p.CWD = strings.TrimSpace(p.CWD)
	p.Provider = normalizeProjectProvider(p.Provider)
	p.ProviderConfig = normalizeProjectProviderConfig(p.ProviderConfig)
	if strings.TrimSpace(p.Name) == "" {
		p.Name = projectNameFromCWD(p.CWD)
	}
	p.Model = normalizeProjectModel(p.Model)
	p.DefaultAgentProfileID = normalizeIdentifier(p.DefaultAgentProfileID)
}

func applyProjectPatchDefaults(p *Project, patch ProjectPatch) ProjectPatch {
	if patch.CWD != nil {
		nextCWD := strings.TrimSpace(*patch.CWD)
		*patch.CWD = nextCWD
		if patch.Name == nil || strings.TrimSpace(*patch.Name) == "" {
			name := projectNameFromCWD(nextCWD)
			patch.Name = &name
		}
	}
	if patch.Name != nil && strings.TrimSpace(*patch.Name) == "" {
		name := projectNameFromCWD(p.CWD)
		patch.Name = &name
	}
	if patch.Model != nil {
		model := normalizeProjectModel(*patch.Model)
		patch.Model = &model
	}
	if patch.Provider != nil {
		provider := normalizeProjectProvider(*patch.Provider)
		patch.Provider = &provider
	}
	if patch.ProviderConfig != nil {
		config := normalizeProjectProviderConfig(*patch.ProviderConfig)
		patch.ProviderConfig = &config
	}
	if patch.DefaultAgentProfileID != nil {
		profileID := normalizeIdentifier(*patch.DefaultAgentProfileID)
		patch.DefaultAgentProfileID = &profileID
	}
	return patch
}

func applyProjectPatch(p *Project, patch ProjectPatch) {
	if patch.Name != nil {
		p.Name = strings.TrimSpace(*patch.Name)
	}
	if patch.CWD != nil {
		p.CWD = strings.TrimSpace(*patch.CWD)
	}
	if patch.Provider != nil {
		p.Provider = normalizeProjectProvider(*patch.Provider)
	}
	if patch.ProviderConfig != nil {
		p.ProviderConfig = normalizeProjectProviderConfig(*patch.ProviderConfig)
	}
	if patch.AutoRun != nil {
		p.AutoRun = *patch.AutoRun
	}
	if patch.Model != nil {
		p.Model = normalizeProjectModel(*patch.Model)
	}
	if patch.ApprovalPolicy != nil {
		p.ApprovalPolicy = *patch.ApprovalPolicy
	}
	if patch.Sandbox != nil {
		p.Sandbox = *patch.Sandbox
	}
	if patch.DefaultAgentProfileID != nil {
		p.DefaultAgentProfileID = normalizeIdentifier(*patch.DefaultAgentProfileID)
	}
}

func (s *Store) nextProjectSortOrder(ctx context.Context) (int, error) {
	var maxOrder int
	err := s.db.QueryRowContext(ctx, `select coalesce(max(sort_order), 0) from projects`).Scan(&maxOrder)
	if err != nil {
		return 0, err
	}
	return maxOrder + 1, nil
}

const projectSelect = `select p.id, p.name, p.cwd, p.provider, p.provider_config_json,
	p.auto_run, p.model, p.approval_policy,
	p.sandbox, p.default_agent_profile_id, p.sort_order, p.created_at, p.updated_at,
	h.reason, h.message, h.hold_since, h.next_check_at, h.last_check_at, h.last_check_error
	from projects p left join project_holds h on h.project_id=p.id`

func projectNameFromCWD(cwd string) string {
	trimmed := strings.TrimRight(strings.TrimSpace(cwd), string(filepath.Separator))
	name := filepath.Base(trimmed)
	if name == "." || name == string(filepath.Separator) || name == "" {
		return "project"
	}
	return name
}

func normalizeProjectModel(model string) string {
	model = strings.TrimSpace(model)
	if model == "" || strings.HasPrefix(strings.ToLower(model), "gemini-") {
		return "codex-default"
	}
	return model
}

func normalizeProjectProvider(provider string) string {
	provider = strings.ToLower(strings.TrimSpace(provider))
	if provider == "" {
		return ProviderCodex
	}
	return provider
}

func normalizeProjectProviderConfig(config string) string {
	config = strings.TrimSpace(config)
	if config == "" {
		return "{}"
	}
	return config
}

func AttachProjectCapabilities(projects []Project) {
	for i := range projects {
		AttachProjectCapability(&projects[i])
	}
}

func AttachProjectCapability(project *Project) {
	project.ProviderCapabilities = ProjectProviderCapabilities(project.Provider)
}

func ProjectProviderCapabilities(provider string) []string {
	capabilities := agent.CapabilitiesForProviderID(provider)
	result := make([]string, 0, len(capabilities))
	for _, capability := range capabilities {
		result = append(result, string(capability))
	}
	return result
}

File: backend-ts/src/db/repositories/projects.ts
import { statSync } from "node:fs";
import type { RunnerDatabase } from "../database.ts";
import { getAgentProfile, type AgentProfile } from "./agentProfiles.ts";
import { normalizeProjectForWrite, normalizeProjectModel, normalizeProjectPatch, normalizeProjectProvider, normalizeProjectProviderConfig, type NormalizedProjectWrite, type ProjectPatchInput, type ProjectWriteInput } from "./projectUtils.ts";
import type { ExecutionPolicyRequest } from "../../providers/core/policyContracts.ts";
import { readStoredExecutionPolicy } from "../../providers/core/policyPersistence.ts";

type ProjectRow = {
  hold_since: unknown;
  last_check_at: unknown;
  last_check_error: unknown;
  message: unknown;
  next_check_at: unknown;
  reason: unknown;
  approval_policy: unknown;
  auto_run: unknown;
  created_at: unknown;
  cwd: unknown;
  default_agent_profile_id: unknown;
  default_mcp_policy_json: unknown;
  default_service_tier: unknown;
  default_skill_policy_json: unknown;
  execution_policy_json: unknown;
  id: unknown;
  model: unknown;
  name: unknown;
  pi_managed: unknown;
  provider: unknown;
  provider_config_json: unknown;
  sandbox: unknown;
  sort_order: unknown;
  updated_at: unknown;
};

export type CreateProjectInput = ProjectWriteInput;

export type UpdateProjectInput = ProjectPatchInput;

export type ProjectHold = { reason: string; message: string; hold_since: string; next_check_at: string; last_check_at: string; last_check_error: string };

export type Project = {
  approval_policy: string;
  auto_run: number;
  created_at: string;
  cwd: string;
  default_agent_profile?: AgentProfile;
  default_agent_profile_id: string;
  default_mcp_policy: string;
  default_service_tier: string;
  default_skill_policy: string;
  execution_policy: ExecutionPolicyRequest;
  execution_policy_json: string;
  execution_policy_source: string;
  execution_policy_warnings: string[];
  hold?: ProjectHold;
  id: string;
  loop_status: string;
  model: string;
  name: string;
  pi_managed: number;
  provider: string;
  provider_capabilities: string[];
  provider_config_json: string;
  sandbox: string;
  sort_order: number;
  updated_at: string;
};

const PROJECT_COLUMNS = `p.id, p.name, p.cwd, p.provider, p.provider_config_json, p.auto_run,
  p.model, p.approval_policy, p.sandbox, p.execution_policy_json, p.default_agent_profile_id, p.default_skill_policy_json,
  p.default_mcp_policy_json, p.default_service_tier, p.sort_order,
  p.created_at, p.updated_at, h.reason, h.message, h.hold_since, h.next_check_at,
  h.last_check_at, h.last_check_error,
  case when exists(select 1 from project_pi_settings s where s.project_id=p.id) then 1 else 0 end as pi_managed`;


export function createProject(db: RunnerDatabase, input: CreateProjectInput): Project {
  const project = normalizeProjectForWrite(input);
  validateProjectForWrite(project.id, project.cwd);
  const existing = getProjectByCWD(db, project.cwd);
  if (existing) return updateProject(db, existing.id, project);
  const timestamp = now();
  const sortOrder = nextProjectSortOrder(db);
  db.sqlite.run(`insert into projects
    (id, name, cwd, provider, provider_config_json, auto_run, model,
     approval_policy, sandbox, execution_policy_json, default_agent_profile_id, default_skill_policy_json, default_mcp_policy_json,
     default_service_tier, sort_order, created_at, updated_at)
    values (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
    [project.id, project.name, project.cwd, project.provider, project.provider_config_json,
      project.auto_run, project.model, project.approval_policy, project.sandbox, project.execution_policy_json,
      project.default_agent_profile_id, project.default_skill_policy, project.default_mcp_policy,
      project.default_service_tier, sortOrder, timestamp, timestamp]);
  return mustGetProject(db, project.id);
}

export function updateProject(db: RunnerDatabase, id: string, input: UpdateProjectInput): Project {
  const projectID = cleanRequiredString(id, "project id");
  const current = getProject(db, projectID);
  if (!current) throw new ProjectNotFoundError();
  const patch = normalizeProjectPatch(projectToWriteShape(current), input);
  const next = { ...projectToWriteShape(current), ...patch };
  validateProjectForWrite(projectID, next.cwd);
  db.sqlite.run(`update projects set name=?, cwd=?, provider=?, provider_config_json=?,
    auto_run=?, model=?, approval_policy=?, sandbox=?, execution_policy_json=?, default_agent_profile_id=?,
    default_skill_policy_json=?, default_mcp_policy_json=?, default_service_tier=?, updated_at=? where id=?`,
    [next.name, next.cwd, next.provider, next.provider_config_json, next.auto_run,
      next.model, next.approval_policy, next.sandbox, next.execution_policy_json, next.default_agent_profile_id,
      next.default_skill_policy, next.default_mcp_policy, next.default_service_tier, now(), projectID]);
  return mustGetProject(db, projectID);
}

export function listProjects(db: RunnerDatabase): Project[] {
  const projects = db.sqlite.query<ProjectRow, []>(`
    select ${PROJECT_COLUMNS} from projects p left join project_holds h on h.project_id=p.id
    order by p.sort_order asc, p.created_at asc, p.id asc
  `).all().map(mapProjectRow);
  return attachAgentProfiles(db, projects);
}

export function getProject(db: RunnerDatabase, id: string): Project | null {
  const projectID = cleanRequiredString(id, "project id");
  const row = db.sqlite.query<ProjectRow, [string]>(`
    select ${PROJECT_COLUMNS} from projects p left join project_holds h on h.project_id=p.id where p.id = ?
  `).get(projectID);
  if (!row) return null;
  return attachAgentProfiles(db, [mapProjectRow(row)])[0] ?? null;
}

function getProjectByCWD(db: RunnerDatabase, cwd: string): Project | null {
  const row = db.sqlite.query<ProjectRow, [string]>(`
    select ${PROJECT_COLUMNS} from projects p left join project_holds h on h.project_id=p.id where p.cwd = ?
  `).get(cwd);
  if (!row) return null;
  return attachAgentProfiles(db, [mapProjectRow(row)])[0] ?? null;
}

function mapProjectRow(row: ProjectRow): Project {
  const provider = normalizeProjectProvider(row.provider);
  const storedPolicy = readStoredExecutionPolicy({
    approvalPolicy: row.approval_policy,
    json: row.execution_policy_json,
    sandbox: row.sandbox,
    scope: "project"
  });
  return {
    id: requiredString(row.id, "projects.id"),
    name: requiredString(row.name, "projects.name"),
    pi_managed: integerValue(row.pi_managed, "projects.pi_managed"),
    cwd: requiredString(row.cwd, "projects.cwd"),
    provider,
    provider_config_json: normalizeProjectProviderConfig(row.provider_config_json),
    auto_run: integerValue(row.auto_run, "projects.auto_run"),
    model: normalizeProjectModel(row.model, provider),
    approval_policy: optionalString(row.approval_policy, "never"),
    sandbox: optionalString(row.sandbox, "workspace-write"),
    execution_policy: storedPolicy.policy!,
    execution_policy_json: optionalString(row.execution_policy_json, "{}"),
    execution_policy_source: storedPolicy.source,
    execution_policy_warnings: storedPolicy.warnings,
    default_agent_profile_id: optionalString(row.default_agent_profile_id),
    default_mcp_policy: optionalString(row.default_mcp_policy_json, "{}"),
    default_service_tier: optionalString(row.default_service_tier),
    default_skill_policy: optionalString(row.default_skill_policy_json, "{}"),
    sort_order: integerValue(row.sort_order, "projects.sort_order"),
    created_at: requiredString(row.created_at, "projects.created_at"),
    updated_at: requiredString(row.updated_at, "projects.updated_at"),
    ...(projectHold(row) ? { hold: projectHold(row) } : {}),
    loop_status: "stopped",
    provider_capabilities: providerCapabilities(row.provider)
  };
}

export class ProjectNotFoundError extends Error {
  constructor() {
    super("资源不存在");
    this.name = "ProjectNotFoundError";
  }
}

function validateProjectForWrite(id: string, cwd: string): void {
  if (id === "") throw new Error("project id 不能为空");
  if (cwd === "") throw new Error("cwd 不能为空");
  validateProjectCWD(cwd);
}

function validateProjectCWD(cwd: string): void {
  try {
    if (!statSync(cwd).isDirectory()) throw new Error("cwd 不是目录");
  } catch (error) {
    if (error instanceof Error && error.message === "cwd 不是目录") throw error;
    throw new Error("cwd 不存在");
  }
}

function mustGetProject(db: RunnerDatabase, id: string): Project {
  const project = getProject(db, id);
  if (!project) throw new ProjectNotFoundError();
  return project;
}

function nextProjectSortOrder(db: RunnerDatabase): number {
  const row = db.sqlite.query<{ value: number }, []>(
    "select coalesce(max(sort_order), 0) + 1 as value from projects"
  ).get();
  return row?.value ?? 1;
}

function projectToWriteShape(project: Project): NormalizedProjectWrite {
  return {
    id: project.id,
    name: project.name,
    cwd: project.cwd,
    provider: project.provider,
    provider_config_json: project.provider_config_json,
    auto_run: project.auto_run,
    model: project.model,
    approval_policy: project.approval_policy,
    sandbox: project.sandbox,
    execution_policy_json: project.execution_policy_json,
    default_agent_profile_id: project.default_agent_profile_id,
    default_service_tier: project.default_service_tier,
    default_mcp_policy: project.default_mcp_policy,
    default_skill_policy: project.default_skill_policy
  };
}

function now(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
}

function cleanRequiredString(value: string, label: string): string {
  const trimmed = value.trim();
  if (trimmed === "") throw new Error(`${label} is required`);
  return trimmed;
}

function requiredString(value: unknown, label: string): string {
  const text = optionalString(value);
  if (text === "") throw new Error(`${label} is required`);
  return text;
}

function optionalString(value: unknown, fallback = ""): string {
  if (value === null || value === undefined) return fallback;
  if (typeof value !== "string") throw new Error(`expected string row value`);
  const trimmed = value.trim();
  return trimmed === "" ? fallback : trimmed;
}

function integerValue(value: unknown, label: string): number {
  if (typeof value !== "number" || !Number.isInteger(value)) throw new Error(`${label} must be an integer`);
  return value;
}

function attachAgentProfiles(db: RunnerDatabase, projects: Project[]): Project[] {
  return projects.map((project) => {
    const profileID = project.default_agent_profile_id.trim();
    if (profileID === "") return project;
    const profile = getAgentProfile(db, profileID);
    return profile ? { ...project, default_agent_profile: profile } : project;
  });
}

function projectHold(row: ProjectRow): ProjectHold | undefined {
  if (optionalString(row.reason) === "") return undefined;
  return {
    reason: optionalString(row.reason),
    message: optionalString(row.message),
    hold_since: optionalString(row.hold_since),
    next_check_at: optionalString(row.next_check_at),
    last_check_at: optionalString(row.last_check_at),
    last_check_error: optionalString(row.last_check_error)
  };
}

function providerCapabilities(value: unknown): string[] {
  switch (normalizeProjectProvider(value)) {
    case "codex":
      return ["issue_execution", "sessions", "resume_session", "interrupt", "approvals", "model_list"];
    case "claude":
      return ["issue_execution", "sessions", "resume_session", "interrupt"];
    case "fake-execution-only":
      return ["issue_execution"];
    default:
      return [];
  }
}
